### PR TITLE
Allow HTML tags in user Markdown input

### LINKED ISSUE #7066: Mardown Inline HTML not working in Wiki
1. **Summary**: The use of raw HTML in Markdown does not work (in the Wiki). This issue was observed in version 6.8.1.

2. **Steps to reproduce**: Create/edit a Wiki page and enter some raw/inline HTML, such as the definition lists described in the Markdown documentation at [github.com][1] or [gitlab.com][2]:
   ```HTML
   <dl>
     <dt>Definition list</dt>
     <dd>Is something people use sometimes.</dd>

     <dt>Markdown in HTML</dt>
     <dd>Does *not* work **very** well. Use HTML <em>tags</em>.</dd>
   </dl>
   ```
   The same problem occurs when using other HTML tags such as a \<table\> tags.

3. **Expected behavior**: The HTML is rendered correctly as seen in the Markdown documentation at [github.com][1] and in the following image:
md-inlinehtml-correct

4. **Observed behavior**: The output is rendered incorrectly as seen in the Markdown documentation at [gitlab.com][2] and in the following image:
md-inlinehtml-incorrect

[1]: https://github.com/gitlabhq/gitlabhq/blob/master/doc/markdown/markdown.md#inline-html
[2]: https://gitlab.com/gitlab-org/gitlab-ce/blob/master/doc/markdown/markdown.md#inline-html

### PR DESCRIPTION
**Pull request by mr-vinn**
This change fixes #7066, and issues [#147](https://gitlab.com/gitlab-org/gitlab-ce/issues/147), [#408](https://gitlab.com/gitlab-org/gitlab-ce/issues/408) and [#476](https://gitlab.com/gitlab-org/gitlab-ce/issues/476) from gitlab.com.

The Markdown parser currently uses Redcarpet's `:filter_html` option, which removes all HTML tags from user input.  The parser also calls ActionView's `#sanitize` method, which removes HTML tags and attributes that are not whitelisted.

GitLab's [Markdown documentation](https://github.com/gitlabhq/gitlabhq/blob/master/doc/markdown/markdown.md#inline-html) says that inline HTML should be allowed, which seems incompatible with the `:filter_html` option.  This change disables that option to allow safe tags from user input to appear in the rendered HTML.

However, relying on `#sanitize` at the end of parsing doesn't handle javascript in tag attributes correctly.  For example, this input:

``` html
<a href="javascript:alert('foo')">link text</a>
```

is rendered as:

``` html
<a href="/namespace101/gitlabhq/blob/markdown/javascript:alert(&#39;foo&#39;)">link text</a>
```

To handle javascript in attributes, I added a call to `#sanitize` before parsing relative links.  This results in the correctly sanitized output:

``` html
<a>link text</a>
```

### CODE REVIEW

**Comment by dblessing** [comment]
ping @randx 

Also, @namnatulco would you be interested in providing documentation in the meantime?

**Comment by namnatulco** [comment]
I hope I'll find some time to do that tonight. Thanks for reminding me!

**Comment by veegee** [comment]
What's the status of this? The gitlab docs show that HTML is possible, but  it doesn't seem to work. This patch sounds like it'll fix that.

**Comment by magikid** [comment]
:+1:

**Comment by lyfeyaj** [comment]
+1 I need this feature!

**Comment by dosire** [comment]
@mr-vinn what is your plan with this?

**Comment by mr-vinn** [comment]
@dosire I'll make the html-pipeline changes that were discussed earlier; I should be able to do that over the weekend.

**Comment by dosire** [comment]
@mr-vinn OK, thanks!

**Comment by mbroz** [comment]
Seems that some special formatting inside table cells is not posible with Markdown (only with added html code).
Also automatic conversion tools (from Google Code wiki for example) generate html tags inside Markdown that is ignored in gitlab (breaking pages).
Please merge support for at least some HTML tags, it increases usability a lot. Thanks!

**Comment by mr-vinn** [comment]
@dosire Sorry for the delay - I added a new change to use the html-pipeline `SanitizationFilter`, so I think this is ready for review.

**Comment by dosire** [comment]
@randx can you review?

File: lib/gitlab/markdown.rb [hunk 1]
```diff
@@ -79,15 +79,34 @@ def gfm_with_options(text, options = {}, project = @project, html_options = {})
...
       #
       # see https://gitlab.com/gitlab-org/html-pipeline-gitlab for more filters
       filters = [
-        HTML::Pipeline::Gitlab::GitlabEmojiFilter
+        HTML::Pipeline::Gitlab::GitlabEmojiFilter,
+        HTML::Pipeline::SanitizationFilter
       ]
 
+      whitelist = HTML::Pipeline::SanitizationFilter::WHITELIST
+      whitelist[:attributes][:all].push('class', 'id', 'style')
```
Comment (randx) [inline lib/gitlab/markdown.rb]: @mr-vinn I am not sure if it makes sense to allow style attributes. I dont want people override GitLab style with custom css

**Comment by randx** [comment]
@mr-vinn please add item to CHANGELOG too

File: lib/gitlab/markdown.rb (hunk 1 above)
Comment (mr-vinn) [inline lib/gitlab/markdown.rb]: @randx That makes sense to me, I'll remove `style` from the attribute whitelist.

**Comment by mr-vinn** [comment]
@randx The CHANGELOG and attribute whitelist have been updated in cc29ce4

**Comment by bbodenmiller** [comment]
@me-vinn looks like you also need a commit squash & debase.

**Comment by mr-vinn** [comment]
@bbodenmiller Thanks for catching that; it should be mergeable now.

**Comment by randx** [comment]
@mr-vinn tests are failing. Can you take a look?

**Comment by mr-vinn** [comment]
@randx I forgot to call `#html_safe` on the parsed output, so Rails was escaping the markup.  I also had to add `<span>` tags to the sanitizer's whitelist to avoid breaking syntax highlighting in code blocks.

**Comment by randx** [comment]
@mr-vinn ping me when it will be ready

**Comment by mr-vinn** [comment]
@randx All set, it's mergeable now.

**Comment by randx** [comment]
@mr-vinn thank you. looks good

**Comment by Razer6** [comment]
@mr-vinn Nice, that you switched to html-pipeline!

**Comment by zlolik** [comment]
Please, specify somewhere, that Ruby should have at least html-pipeline-2.0 in order to work according to documentation:
```html
<details><summary>&#9733; About html-pipeline-2.0</summary>Only that version have [merge 171 ](https://github.com/jch/html-pipeline/pull/171) that fix `<summary>` and `<details>` tags.
</details>
```
<details><summary>&#9733; About html-pipeline-2.0</summary>Only that version have [merge 171 ](https://github.com/jch/html-pipeline/pull/171) that fix `<summary>` and `<details>` tags.
</details>

**Pull request merged**